### DISCUSSION TITLE
Self referencing objects in Optional

### DISCUSSION THREAD

**Discussion by dineshkumar02** [danielaparker/jsoncons#292] [General] [answered]
Hi Team,

I am trying to use the self referencing objects in the optional, but I am getting error like below.

~\workspace\jsoncons\include\jsoncons\detail\optional.hpp(53,15): error C2079: 'jsoncons::detail::optional<ns::hiking_reputon>::value_' uses undefined class 'ns::hiking_reputon'

And, the class I have declared as below.

   ```
class hiking_reputon;
class hiking_reputon
    {
        std::string rater_;
        hiking_experience assertion_;
        std::string rated_;
-------
        jsoncons::optional<hiking_reputon> reputon_;
-------
```
Do, please let me know if I am doing any wrong here.

Thanks in advance for your time and help.

--Dinesh

**Comment by danielaparker** [comment, marked as answer]
That wouldn't work with `std::optional` either (`jsoncons::optional` is typedefed to `std::optional` if C++17 is detected.) Whether `jsoncons` or `std`, `optional<T>` requires a complete type, it stores a `T` _within_ the optional object, not on the heap.

For self referencing objects, I would suggest using `std::unique_ptr` instead.

**Reply by dineshkumar02** [reply]
Thank you very much @danielaparker. Will try your suggestions

**Reply by dineshkumar02** [reply]
@danielaparker , thank you for your suggestion regarding the unique_ptr.

I have tried a lot using the unique_ptr within in the class, and was hitting the hard limits of unique_ptr like, we can't copy the unique_ptr object. And after going through the documentation, I've found an example where unique_ptr is used in struct rather class.

I took that documentation as reference and able to get the self reference parsing.
I am adding a working example here, which will help others as well.
```

#include <jsoncons/json.hpp>
#include <jsoncons/decode_json.hpp>
#include <jsoncons/encode_json.hpp>
#include <jsoncons_ext/jsonpath/jsonpath.hpp>
#include <iostream>
using namespace jsoncons;

namespace ns {
    struct smart_pointer_test
    {
        std::string rater;
        std::string rated;
        double rating;
        std::unique_ptr<smart_pointer_test> self;
    };

} // namespace ns

JSONCONS_N_MEMBER_TRAITS(ns::smart_pointer_test, 3, rater, rated, rating, self)

int main()
{
    std::string data = R"(
    {
           "rater": "HikingAsylum",         
           "rated": "Marilyn C",
           "rating": 0.90,
           "self": 
            {
               "rater": "HikingAsylum",
               "rated": "Marilyn C",
               "rating": 0.90,
               "self": {
                "rater": "HikingAsylum",
                "rated": "Marilyn C",
                "rating": 0.90
                }
            }         
    }
  )";
    std::string s;
 
    try {
        ns::smart_pointer_test v = jsoncons::decode_json<ns::smart_pointer_test>(data);
        jsoncons::encode_json_pretty(v, s);
        std::cout << "self reference parsed string is " << std::endl << s << std::endl;
    }
    catch (std::exception& e) {
        std::cerr << "got error while parsing json. error: " << e.what() << std::endl;
    }
}

```

**Output**
```
self reference parsed string is
{
    "rated": "Marilyn C",
    "rater": "HikingAsylum",
    "rating": 0.9,
    "self": {
        "rated": "Marilyn C",
        "rater": "HikingAsylum",
        "rating": 0.9,
        "self": {
            "rated": "Marilyn C",
            "rater": "HikingAsylum",
            "rating": 0.9
        }
    }
}
```
